opensafely/unzip.py

```diff
@@ -0,0 +1,44 @@
+import gzip
+import sys
+from pathlib import Path
+
+DESCRIPTION = "Unzip a compressesd .csv.gz or .dta.gz file for local viewing"
+
+
+def add_arguments(parser):
+    parser.add_argument(
+        "path", help="path to unzip. If a directory, unzip all .gz files in it"
+    )
+
+
+def main(path):
+    p = Path(path)
+    if p.is_dir():
+        paths = p.glob("**/*.gz")
+    else:
+        if not p.exists():
+            sys.exit(f"{p} does not exist")
+        paths = [p]
+
+    for p in paths:
+        uncompressed = p.with_suffix("")  # strip .gz
+        if uncompressed.exists():
+            if uncompressed.stat().st_mtime > p.stat().st_mtime:
+                print(f"{p} already unzipped to {uncompressed}")
+                continue
+
+        unzip(p, uncompressed)
+        print(f"{p} unzipped to {uncompressed}")
+
+
+def unzip(src_path, dst_path):
+    try:
+        tmp = dst_path.with_suffix(".tmp")
+        with tmp.open("wb") as dst:
+            with gzip.open(src_path) as src:
+                while block := src.read(8192):
+                    dst.write(block)
+    except Exception:
+        tmp.unlink(missing_ok=True)
+    else:
+        tmp.rename(dst_path)
```

Comment:
I think `replace()` is better behaved on Windows, although I can't quite remember the details:
https://docs.python.org/3.8/library/pathlib.html#pathlib.Path.replace

Reply:
TIL: https://stackoverflow.com/questions/60369291/os-link-vs-os-rename-vs-os-replace-for-writing-atomic-write-files-what/65553754#answer-65553754